bpf/bpf_lxc.c

```diff
@@ -1644,6 +1681,61 @@ int handle_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+/* Handle policy decisions as the packet makes its way from the
+ * endpoint.  Previously, the packet has come from the same endpoint,
+ * but was redirected to a L7 LB.
+ *
+ * This program will be tail called from bpf_host for packets sent by
+ * a L7 LB.
+ */
+__section_tail(CILIUM_MAP_EGRESSPOLICY, TEMPLATE_LXC_ID)
+int handle_policy_egress(struct __ctx_buff *ctx)
+{
+	__u16 proto;
+	int ret;
+	int ifindex;
+
+	if (!validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
+		goto out;
+	}
+
+	ifindex = ctx_load_meta(ctx, CB_IFINDEX);
+	ctx_store_meta(ctx, CB_IFINDEX, 0); /* Check if this is needed? */
+	ctx_store_meta(ctx, CB_FROM_HOST, FROM_HOST_L7_LB);
+
+	edt_set_aggregate(ctx, 0); /* do not count this traffic again */
+	send_trace_notify(ctx, TRACE_FROM_PROXY, SECLABEL, 0, 0,
+			  ifindex, 0, TRACE_PAYLOAD_LEN);
+
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		invoke_tailcall_if(__or(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
```

Comment:
why do we only invoke it when both IPv4 and IPv6 are enabled? Also we know for sure IPv6 is enabled here.

Reply:
that is some tail_call/complexity optimization logic I have copy/pasted from elsewhere in the file. My reading is that the function will be tail-called into ONLY if both IPv4 and IPv6 are defined, or DEBUG is defined, otherwise the function will be inlined (without a tail call).

IMO we should keep this form intact and not optimize it for the current context (where ENABLE_IPV6 is defined).

Reply:
Ah I see, thanks!